src/System.Security.Cryptography.Algorithms/tests/AesCcmTests.cs

```diff
@@ -0,0 +1,433 @@
+                byte[] ciphertext = new byte[dataLength];
+                byte[] key = new byte[16];
+                byte[] nonce = new byte[AesCcm.NonceByteSizes.MinSize];
+                byte[] tag = new byte[AesCcm.TagByteSizes.MinSize];
+                rng.GetBytes(key);
+                rng.GetBytes(nonce);
+
+                using (AesCcm aesCcm = new AesCcm(key))
+                {
+                    aesCcm.Encrypt(nonce, plaintext, ciphertext, tag, additionalData);
+
+                    byte[] decrypted = new byte[dataLength];
+                    aesCcm.Decrypt(nonce, ciphertext, tag, decrypted, additionalData);
+                    Assert.Equal(plaintext, decrypted);
+                }
+            }
+        }
+
+        [Theory]
+        [InlineData(0, 1)]
+        [InlineData(0, 30)]
+        [InlineData(1, 1)]
+        [InlineData(1, 100)]
+        [InlineData(7, 12)]
+        [InlineData(16, 16)]
+        [InlineData(17, 29)]
+        [InlineData(32, 7)]
+        [InlineData(41, 25)]
+        [InlineData(48, 22)]
+        [InlineData(50, 5)]
+        public static void EncryptTamperAADDecrypt(int dataLength, int additionalDataLength)
+        {
+            Random r = new Random();
+            using (var rng = RandomNumberGenerator.Create())
+            {
+                byte[] additionalData = new byte[additionalDataLength];
+                rng.GetBytes(additionalData);
+
+                byte[] plaintext = Enumerable.Range(1, dataLength).Select((x) => (byte)x).ToArray();
+                byte[] ciphertext = new byte[dataLength];
+                byte[] key = new byte[16];
+                byte[] nonce = new byte[AesCcm.NonceByteSizes.MinSize];
+                byte[] tag = new byte[AesCcm.TagByteSizes.MinSize];
+                rng.GetBytes(key);
+                rng.GetBytes(nonce);
+
+                using (AesCcm aesCcm = new AesCcm(key))
+                {
+                    aesCcm.Encrypt(nonce, plaintext, ciphertext, tag, additionalData);
+
+                    int tamperedIdx = r.Next() % additionalData.Length;
+                    additionalData[tamperedIdx] = (byte)(additionalData[tamperedIdx] + 1);
+
+                    byte[] decrypted = new byte[dataLength];
+                    Assert.Throws<CryptographicException>(() => aesCcm.Decrypt(nonce, ciphertext, tag, decrypted, additionalData));
+                }
+            }
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData(1)]
+        [InlineData(2)]
+        [InlineData(17)]
+        [InlineData(29)]
+        [InlineData(33)]
+        public static void InvalidKeyLength(int keyLength)
+        {
+            using (var rng = RandomNumberGenerator.Create())
+            {
+                byte[] key = new byte[keyLength];
+                rng.GetBytes(key);
+
+                Assert.Throws<CryptographicException>(() => new AesCcm(key));
```

Comment:
Since this is "your input is bad" it should be something based on ArgumentException